Lab-2/leoenge_lab2.ipynb

```diff
@@ -0,0 +1,916 @@
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "LYKf4XgS92qG",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def backward_substitution(matrix, vector):\n",
+        "  if not (type(matrix) == np.ndarray and matrix.ndim == 2 and np.linalg.det(matrix) != 0):\n",
+        "    raise Exception(\"The input matrix is bad...\")\n",
+        "  if not (type(vector) == np.ndarray and vector.ndim == 1):\n",
+        "    raise Exception(\"The input vector is bad...\")\n",
+        "  if not(matrix.shape[0] == vector.size):\n",
+        "    raise Exception(\"The dimensions of the vector and the matrix are not compatible.\")\n",
+        "  \n",
+        "  n = vector.size\n",
+        "  result = np.zeros(n)\n",
+        "  result[-1] = vector[-1]/matrix[-1,-1]\n",
+        "  for i in range(n-2, -1, -1):\n",
+        "    sum = 0\n",
+        "    for j in range(i+1, n):\n",
+        "      sum += matrix[i,j]*result[j]\n",
+        "    result[i] = (vector[i] - sum)/matrix[i,i]\n",
+        "  return result\n"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "ysXWDdylG6Lx",
+        "colab_type": "code",
+        "colab": {}
+      },
+      "source": [
+        "def equation_system_solver(matrix, vector):\n",
+        "  if not (type(matrix) == np.ndarray and matrix.ndim == 2 and np.linalg.det(matrix) != 0):\n",
+        "    raise Exception(\"The input matrix is bad...\")\n",
+        "  if not (type(vector) == np.ndarray and vector.ndim == 1):\n",
+        "    raise Exception(\"The input vector is bad...\")\n",
+        "  if not(matrix.shape[0] == vector.size):\n",
+        "    raise Exception(\"The dimensions of the vector and the matrix are not compatible.\")\n",
+        "  \n",
+        "  n = vector.size\n",
+        "  Q, R = qr_factorization(matrix)\n",
+        "  return backward_substitution(R, Q.transpose().dot(vector))"
+      ],
+      "execution_count": 0,
+      "outputs": []
+    },
+    {
+      "cell_type": "markdown",
+      "metadata": {
+        "id": "rfchOQdmNAlI",
+        "colab_type": "text"
+      },
+      "source": [
+        "### Testing Equation System Solver\n",
+        "The equation solver is tested both by calculating the residual $\\| Ax-b \\|$ and also by calculating the \"reverse\" residual, which is given by $\\| y-x\\|$ where $y$ is an arbitrary vector and $x$ is the solution to the system $Ax = Ay$. "
+      ]
+    },
+    {
+      "cell_type": "code",
+      "metadata": {
+        "id": "K3UqXBi1Gquh",
+        "colab_type": "code",
+        "outputId": "a351a589-f45e-4149-83eb-06799f304f8d",
+        "colab": {
+          "base_uri": "https://localhost:8080/",
+          "height": 238
+        }
+      },
+      "source": [
+        "class Test(unittest.TestCase):\n",
+        "  def test_singular(self):\n",
```

Comment:
This test seems to fail no matter what changes I make. Look through it a second time to confirm it behaves the way you want it to and make updates accordingly.